hazelcast/src/main/java/com/hazelcast/util/ExceptionUtil.java

```diff
@@ -48,23 +48,39 @@ public static String toString(Throwable cause) {
         return sw.toString();
     }
 
+    public static RuntimeException peel(final Throwable t) {
+        return (RuntimeException) peel(t, null);
+    }
+
+    public static <T extends Throwable> Throwable peel(final Throwable t, Class<T> allowedType) {
+        if (t instanceof RuntimeException) {
+            return t;
+        }
+
+        if (t instanceof ExecutionException) {
+            final Throwable cause = t.getCause();
+            if (cause != null) {
+                return peel(cause, allowedType);
+            } else {
+                return new HazelcastException(t);
```

Comment:
@sancar please fix these bugs and add tests. Also add tests for dealing with interrupt status.